### PR TITLE
Add option to specify/broadcast RNN init hidden state

### PR DESCRIPTION
**Pull request by sallamander**
This PR builds in the option to use the final hidden state of an RNN encoder as the initial hidden state of the RNN decoder. This is done in each of the following papers - [Sequence to Sequence Learning with Neural Networks](https://arxiv.org/pdf/1409.3215v3.pdf), [Learning Phrase Representations using RNN Encoder-Decoder for Statistical Machine Translation](https://arxiv.org/pdf/1406.1078v3.pdf), and [Neural Machine Translation by Jointly Learning to Align and Translate](https://arxiv.org/pdf/1409.0473v7.pdf). There are also a couple of issues revolving around it - https://github.com/fchollet/keras/issues/2995, https://github.com/fchollet/keras/issues/2837, and https://github.com/fchollet/keras/issues/694. 

It's worth noting that I didn't throw a `tanh` around the final hidden state when transferring it to the decoder (like the last two of those papers), but that could be easily added. Other techniques for taking the final hidden state(s) of the encoder and using them as the initial hidden state could also easily be added (for example, taking the mean).

### CODE REVIEW

**Comment by RamaneekGill** [comment]
@sallamander @fchollet any issues with this PR other than resolving conflicts?

**Comment by sallamander** [comment]
I was unaware that there were issues @RamaneekGill - thanks for bringing that to my attention. @fchollet If this is something that would be useful, I'm happy to clean it up and resolve the conflicts!

**Comment by sallamander** [comment]
@RamaneekGill There were some changes to the TF RNN backend such that the number of timesteps is no longer in the output of the cell anymore (it's simply `None`). So, when I try to grab it [here](https://github.com/fchollet/keras/pull/3947/files#diff-a18b8c6a1191d6f49303e0d599ca8c37R2989), we run into issues.

I've traced it back to the change from using `_dynamic_rnn_loop` to the use of `control_flow_ops.while_loop` in [this commit](https://github.com/fchollet/keras/commit/a9b6bef0624c67d6df1618ca63d8e8141b0df4d0#diff-a18b8c6a1191d6f49303e0d599ca8c37R1294). This is something that could probably be fixed in the backend file itself, or we could simply force users to specify the number of timesteps in a similar way that they have to when using `stateful=True`. @fchollet Any thoughts here?

**Comment by sallamander** [comment]
Mentioning https://github.com/fchollet/keras/pull/2991, which had the same goal in mind. It'd be really nice to get this added to Keras.

**Comment by RamaneekGill** [comment]
As a temporary workaround having explicit declarations of the time step is adequate enough in my opinion. It's used for when making RNN cells stateful for I believe remembering the hidden state between batches. 

Personally I use stateful training quite often so I'm not opposed to this but implementation wise it's best to keep an open issue if this gets merged. We'll see @fchollet's thoughts when the code review is done. 

BTW are the docs autogenerated? If not a documentation update should accompany this PR especially since RNNs are used quite often and initializing hidden states is necessary for many sequence to sequence learning specifications.

**Comment by sallamander** [comment]
@RamaneekGill Alright, I've made the change so it now uses the `input_shape`. As far as I know, the docs are auto-generated, yes. @fchollet this is ready for review 😁

**Comment by harikrishnavydana** [comment]
1.How do we access the above version of keras which as the changes mentioned in the above ..
2.does this have a provision to connect output of a time step to use as input to next time step?

**Comment by RamaneekGill** [comment]
@harikrishnavydana this PR introduces hidden state broadcasting and initialization for RNN layers. With broadcasting it will automatically take the last output of the previous layer and initialize the current layer's hidden state with it, note that the input shape along with batch size has to be identified here.

With the hidden state init option you can pass in any numpy array to set the current layer's hidden state.

@sallamander what happens if the previous layer isn't an RNN layer? It seems like this PR will use the output and still initialize it seems like the tests only uses RNN cells. I can see a future use case where someone might want to initialize with convolution filters, will a situation like this break the implementation without raising an appropriate error?

**Comment by RamaneekGill** [comment]
@sallamander looking over the contributing guide it suggested to rebase and squash commits to keep the commit history of the repo clean. http://rebaseandsqua.sh/ might be worth doing before the review process begins

**Comment by sallamander** [comment]
@RamaneekGill That's a good question, and something I've left to a future PR (or to address in review if we think necessary). Right now this is built to accept the output of a previous RRN layer as input because it assumes that the time axis is at 1, but in theory *I think* it would be as easy as allowing the user to pass in any input and specify the axes that need to be pulled from that input for initialization. I'll put an error in there for now... good call 👍 

I had avoided the rebase given the length of time that passed since the PR was opened, but you're right that I should do it. I'll plan on doing sometime today. Thanks for the review!

**Comment by harikrishnavydana** [comment]
As mentioned by @sallamander in pr request  using the op of last layer RNN for initializint the next RNN hidden layer....
2.along with that using the output of previous time step as input of the present time step (the Depth first implementation of RNN) is also a crucial aspect in these papers. may be incorporating it to keras RNN would be much useful @RamaneekGill

**Comment by harikrishnavydana** [comment]
@sallamander @fchollet @RamaneekGill 

i have tried to use broadcasting option to build sequence to sequence.....but i get the gradient problems. can u figure out some issue


model=Sequential()

input=Input(batch_shape=(1,None,n_in))
td1=TimeDistributed(Dense(10, activation='softmax'))(input)
l1=SimpleRNN(n_hidden, activation='relu',return_sequences=False)(td1)
l2=RepeatVector(1)(l1)
for i in range(80):
	if i==0:
		#M1= merge([l2,l2], name="merged"+str(i), mode='concat')
		l3=SimpleRNN(n_hidden, activation='relu',return_sequences=True,hidden_state_broadcasted=True,name='decoder-first'+str(i))(l2)
		out=l3
	else:
		#print i
		M1= merge([l2,l3], name="merged"+str(i), mode='concat')	
		l3=SimpleRNN(n_hidden, activation='relu',return_sequences=True,hidden_state_broadcasted=True,name='decoder-first'+str(i))(M1)
		out = Lambda(append_outputs, output_shape=(i+1,n_hidden,), name="h_n"+str(i))([out,l3])
sf=TimeDistributed(Dense(52, activation='softmax'))(out)
	#output=[]
model = Model(input=input, output=[sf])
adadelta=Adadelta(lr=0.0000001, rho=0.95, epsilon=1e-06)
model.compile(loss='categorical_crossentropy',optimizer=adadelta, metrics=['accuracy'])





error:  File "RNN_broadcasting.py", line 125, in <module>
    model.fit(next(tr)[0], next(tr)[1], nb_epoch=1,verbose=1)
  File "/home/filter/harikrishna/keras_with_pelu_ncelu/broadcasting_RNN/keras/keras/engine/training.py", line 1169, in fit
    self._make_train_function()
  File "/home/filter/harikrishna/keras_with_pelu_ncelu/broadcasting_RNN/keras/keras/engine/training.py", line 761, in _make_train_function
    self.total_loss)
  File "/home/filter/harikrishna/keras_with_pelu_ncelu/broadcasting_RNN/keras/keras/optimizers.py", line 347, in get_updates
    grads = self.get_gradients(loss, params)
  File "/home/filter/harikrishna/keras_with_pelu_ncelu/broadcasting_RNN/keras/keras/optimizers.py", line 82, in get_gradients
    grads = K.gradients(loss, params)
  File "/home/filter/harikrishna/keras_with_pelu_ncelu/broadcasting_RNN/keras/keras/backend/theano_backend.py", line 973, in gradients
    return T.grad(loss, variables)
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 553, in grad
    grad_dict, wrt, cost_name)
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 1317, in _populate_grad_dict
    rval = [access_grad_cache(elem) for elem in wrt]
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 1272, in access_grad_cach
"/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 965, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 1272, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 965, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 1272, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 965, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 1272, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 965, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/usr/local/lib/python2.7/dist-packages/theano/gradient.py", line 1272, in access_grad_cache

**Comment by sallamander** [comment]
@harikrishnavydana Can you link to a gist or otherwise clean version of your code, please?

**Comment by harikrishnavydana** [comment]
@sallamander for every time step i am using a broadcast RNN. so that the output of that RNN can be used as input to next time step. i am manually iterating these broadcast RNNS over required number of time steps that

**Comment by sallamander** [comment]
@harikrishnavydana Yeah, that makes sense. To help fix or debug this, though, I need a link or a copied/pasted version of the code. I can't figure out what's going wrong from just the stack trace above...

**Comment by harikrishnavydana** [comment]
[code.txt](https://github.com/fchollet/keras/files/776080/code.txt)

**Comment by sallamander** [comment]
@harikrishnavydana Can you please send the full version of the code that generated that stack trace? The code you provided has missing import statements, missing variables, and after fixing those, there are reshape errors from the Lambda layer. Rather than trying to guess exactly what you want to do, it would be easiest (and probably most beneficial for you) if you could send the full script that I can run to regenerate this error.

**Comment by harikrishnavydana** [comment]
[debuging.tar.gz](https://github.com/fchollet/keras/files/776114/debuging.tar.gz)
[debuging.zip](https://github.com/fchollet/keras/files/776115/debuging.zip)
@sallamander this has the code i was running with data folder having ten files

**Comment by sallamander** [comment]
@harikrishnavydana I've gotten this to mostly work, but you've referenced a variable that has not been initialized, and then even after that the way you are merging in the Lambda layers is (I'm pretty sure) going to cause reshape errors when you iterate. I don't think this is an issue with the new broadcasting code, but rather your implementation. We should move this offline if it's going to continue - feel free to DM me if you want to chat more. 

http://pastebin.com/bUGyyi31

Note: I'm running on Python 3, with `tensorflow 0.12`.

**Comment by sallamander** [comment]
ping @fchollet any feedback on this?

**Comment by harikrishnavydana** [comment]
@sallamander i have used python 2.7 with theano backend

**Comment by harikrishnavydana** [comment]
@sallamander  can u send ur gamil account .. i would like to discuss more

**Comment by sallamander** [comment]
sean@arterys.com

**Comment by Gliganu** [comment]
@sallamander 

can you provide an approximate timeline for when/if this PR will be merged ?

**Comment by sallamander** [comment]
@Gliganu It's not up to me to merge it, so I can't give an approximate timeline. Sorry!

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.
